packages/core/src/types.ts

```diff
@@ -74,6 +91,42 @@ export type Action<TContext, TEvent extends EventObject> =
   | ActionObject<TContext, TEvent>
   | ActionFunction<TContext, TEvent>;
 
+/**
+ * Extracts action objects that have no extra properties.
+ */
+type SimpleActionsFrom<T extends BaseActionObject> = ActionObject<
+  any,
+  any
+> extends T
+  ? T // If actions are unspecified, all action types are allowed (unsafe)
+  : T extends {
+      [K in keyof Omit<T, 'type'>]: never;
+    }
+  ? T
```

Comment:
This condition is slightly indirect. It maps all other properties to `never` and then checks if `T` is a subtype of that, where it won't be if any key has been successfully remapped. The truthy branch is chosen when we end up with `T extends {}`.

A more straightforward approach to this problem could like this:
```ts
type ExtractSimple<T> = T extends any
  ? Exclude<keyof T, "type"> extends never
    ? T
    : never
  : never;

/**
 * Extracts action objects that have no extra properties.
 */
type SimpleActionsFrom<T extends BaseActionObject> = ActionObject<
  any,
  any
> extends T
  ? T // If actions are unspecified, all action types are allowed (unsafe)
  : ExtractSimple<T>;
```

This seems to test if the union member has only the `type` property more explicitly.

Note that we already have a slight variation of this in this very file:
https://github.com/statelyai/xstate/blob/58fd61d7c4f752dd517a25616b9a47dc39323fe2/packages/core/src/types.ts#L261-L265

Reply:
Sure, feel free to refactor this if the outcome is the same.

Reply:
(Just for the sake of bike-shedding & perhaps micro-blogging :P)

The most principally correct way of writing `ExtractSimple` is this...
```ts
type ExtractSimple<T> = T extends any ? { type: Prop<T, "type"> } extends T ? T : never : never

// for posterity
type Prop<T, K> = K extends keyof T ? T[K] : never
```
What we ultimately want to check is -- is `{ type: action.type }` assignable to `action` and that's exactly and literally what we're doing when we write `Pick<T, "type"> extends T ? T`.

All other ways of writing it are, if I may, incorrect. Even if they produce same results. And fun fact -- they don't. For example in Matuesz's `ExtractSimple` what he's doing is "check if there are no keys other than `type`" which is not what we want to check -- for `ExtractSimple<{ type: "foo", bar?: number }>` his version would produce `never` and my version would produce `"foo"` and mine is correct (unless the original requirement is not about assignability)

The reason I often champion to be "principally correct" rather than say "utilitarian correct" is that you don't have to think about edge cases. For example if didn't write my version thinking about the optional edge case, I just wrote it the way it was supposed to be written and then I just happen to realize this downside in Matuesz's version.

But why do I say "Even if they produce same results"? For being future proof. What if tomorrow typescript introduces a modifier say "unqiue" so that `keyof { unqiue type: "foo" }` is `"type"` but `{ type: "foo" }` is not assignable to `{ unqiue type: "foo" }` (unqiue essentially meaning that property value can't and shouldn't be produced alone) -- in that case my version would still work and some other version might fail.

Another good take away is that we aren't used to leveraging the fact that `extends` checks assignability and come up with really obscure ways to check it, like [Steve was doing and I suggested a refactored](https://twitter.com/devanshj__/status/1416468275408478209) and even the versions here are somewhat indirections.

My sincere apologies to use this space as my blog (high time I make one) and please ignore this comment -- use any version it doesn't matter much :P

Reply:
I actually thought about the optional properties case myself - just didn't figure out how to support it and put it aside as potential future improvement.

Thanks for providing the variant handling this!

Reply:
Ah gotcha. You're welcome!